AsyncDisplayKit/ASDisplayNode.mm

```diff
@@ -553,8 +551,13 @@ - (void)_loadViewOrLayerIsLayerBacked:(BOOL)isLayerBacked
   if (isLayerBacked) {
     TIME_SCOPED(_debugTimeToCreateView);
     _layer = [self _layerToLoad];
-    // Surpress warning for Base SDK > 10.0
-    _layer.delegate = (id<CALayerDelegate>)self;
+    static int ASLayerDelegateAssociationKey;
+
+    // We cannot make self the layer's delegate directly, because the delegate is
+    // actually `assign` though the docs say weak.
+    ASWeakProxy *instance = [ASWeakProxy weakProxyWithTarget:self];
+    _layer.delegate = (id<CALayerDelegate>)instance;
+    objc_setAssociatedObject(_layer, &ASLayerDelegateAssociationKey, instance, OBJC_ASSOCIATION_RETAIN_NONATOMIC);
```

Comment:
Hmm, I didn't realize we would need this, although it makes sense.  We should take this diff anyway, but I'm pretty confident this will show up in profiling.  I can think of two costs:

- Object allocation / deallocation of ASWeakProxy (not just total time, but objc runtime locks / malloc contention when the system is running at load)
- Associated object definition and also teardown when layer is deallocated
- All traffic to the delegate needs to access the associated object too, which also includes retain/release traffic
- Finally, proxy-related overhead for checking selectors...

Probably the key observation is that almost all of this work will be confined to the main thread (although still cause micro-contentions with background threads).  layoutSublayers, display, etc.

It seems to me like setting _layer.delegate = _layer is more utilitarian.  I believe we already have a nodeDelegate or the layer just accesses the node directly anyway, so the code should be short.

All that said, we may not use the delegate as much as it could be used.  I think we override -layoutSublayers and -display directly instead of using the delegate method, for example.  So maybe the traffic is very low.